File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/exception/CircularDependencyException.java
package ru.tinkoff.kora.kora.app.annotation.processor.exception;

import com.palantir.javapoet.TypeName;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.annotation.processor.common.ProcessingError;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;
import ru.tinkoff.kora.kora.app.annotation.processor.declaration.ComponentDeclaration;

import java.util.List;
import java.util.stream.Collectors;

public class CircularDependencyException extends ProcessingErrorException {

    public CircularDependencyException(List<ComponentDeclaration> cycle, ComponentDeclaration declaration) {
        super(getError(cycle, declaration));
    }

    private static ProcessingError getError(List<ComponentDeclaration> cycle,
                                            ComponentDeclaration declaration) {
        var deps = cycle.stream()
            .map(c -> String.format("- %s", c.declarationString()))
            .collect(Collectors.joining("\n", "Cycle dependency candidates:\n", "")).indent(2);

        if (declaration.tags().isEmpty()) {
            return new ProcessingError("Encountered circular dependency in graph for source type: " + TypeName.get(declaration.type()) + " (no tags)\n"
                                       + deps
                                       + "\nPlease check that you are not using cycle dependency in %s, this is forbidden.".formatted(CommonClassNames.lifecycle),
                declaration.source());
        } else {
            var tagMsg = declaration.tags().stream()
                .collect(Collectors.joining(", ", "@Tag(", ")"));
            return new ProcessingError("Encountered circular dependency in graph for source type: " + TypeName.get(declaration.type()) + " with " + tagMsg + "\n"
                                       + deps
                                       + "\nPlease check that you are not using cycle dependency in %s, this is forbidden.".formatted(CommonClassNames.lifecycle),
                declaration.source());
        }
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/exception/CircularDependencyException.kt
package ru.tinkoff.kora.kora.app.ksp.exception

import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.kora.app.ksp.declaration.ComponentDeclaration
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.exception.ProcessingError
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import java.util.stream.Collectors

data class CircularDependencyException(
    val cycle: List<ComponentDeclaration>,
    val declaration: ComponentDeclaration
) : ProcessingErrorException(
    getError(cycle, declaration)
) {

    companion object {
        private fun getError(
            cycle: List<ComponentDeclaration>,
            declaration: ComponentDeclaration
        ): ProcessingError {
            val deps = cycle
                .map { String.format("- %s", it.declarationString()) }
                .joinToString("\n", "Cycle dependency candidates:\n", "").prependIndent("  ")

            if (declaration.tags.isEmpty()) {
                return ProcessingError(
                    """Encountered circular dependency in graph for source type: ${declaration.type.toTypeName()} (no tags)
                    $deps
                    Please check that you are not using cycle dependency in ${CommonClassNames.lifecycle}, this is forbidden.""".trimIndent(),
                    declaration.source
                )
            } else {
                val tagMsg: String = declaration.tags.stream()
                    .collect(Collectors.joining(", ", "@Tag(", ")"))
                return ProcessingError(
                    """Encountered circular dependency in graph for source type: ${declaration.type.toTypeName()} with $tagMsg
                    $deps
                    Please check that you are not using cycle dependency in ${CommonClassNames.lifecycle}, this is forbidden.""".trimIndent(),
                    declaration.source
                )
            }
        }
    }
}
